src/mapreduce/schedule.go

```diff
@@ -14,15 +14,28 @@ func (mr *Master) schedule(phase jobPhase) {
 		ntasks = mr.nReduce
 		nios = len(mr.files)
 	}
-
 	fmt.Printf("Schedule: %v %v tasks (%d I/Os)\n", ntasks, phase, nios)
+	// make sure a worker is dispatched for ever ntasks
+
+	var done = make(chan bool)
+
+	for i := 0; i < ntasks; i++ {
+		go func(i int) {
+			worker := <-mr.registerChannel
+			args := DoTaskArgs{mr.jobName, mr.files[i], phase, i, nios}
+			ok := call(worker, "Worker.DoTask", args, new(struct{}))
+			for ok != true {
+				worker = <-mr.registerChannel
+				ok = call(worker, "Worker.DoTask", args, new(struct{}))
+			}
+			done <- true
+			mr.registerChannel <- worker
+		}(i)
+	}
+
+	for i := 0; i < ntasks; i++ {
+		<-done
+	}
```

Comment:
This is the concurrent heart of the program. It pulls workers of the worker channel and dispatches work using an RPC call. If the worker is good, they're put back into the pool at line 32. If the RPC call fails, we consider the worker bad and dump them, pulling a new worker off the worker channel.
After all is said and done, we wait for the scheduled number of tasks to complete.